gift-certificates-app/dao/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java

```diff
@@ -1,94 +1,169 @@
-    private final CertificateGetSQLRequest getSQLRequest;
-    private final CertificateUpdateSQLRequest updateSQLRequest;
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    private final TagDAO tagDAO;
 
     @Autowired
-    public CertificateDAOImpl(JdbcTemplate jdbcTemplate, CertificateGetSQLRequest getSQLRequest,
-                              CertificateUpdateSQLRequest updateSQLRequest) {
-        this.jdbcTemplate = jdbcTemplate;
-        this.getSQLRequest = getSQLRequest;
-        this.updateSQLRequest = updateSQLRequest;
+    public CertificateDAOImpl(TagDAO tagDAO) {
+        this.tagDAO = tagDAO;
     }
 
     @Override
-    public void add(Certificate certificate) {
-        certificate.setCreateDate(Instant.now().truncatedTo(ChronoUnit.MILLIS));
-        certificate.setLastUpdateDate(Instant.now().truncatedTo(ChronoUnit.MILLIS));
-        KeyHolder keyHolder = new GeneratedKeyHolder();
-        jdbcTemplate.update(
-                connection -> {
-                    PreparedStatement ps =
-                            connection.prepareStatement(CREATE_CERTIFICATE, new String[]{"id"});
-                    ps.setString(1, certificate.getName());
-                    ps.setString(2, certificate.getDescription());
-                    ps.setInt(3, certificate.getDuration());
-                    ps.setDouble(4, certificate.getPrice());
-                    ps.setTimestamp(5, Timestamp.from(certificate.getCreateDate()));
-                    ps.setTimestamp(6, Timestamp.from(certificate.getLastUpdateDate()));
-                    return ps;
-                },
-                keyHolder);
-
-        if (keyHolder.getKey() != null) {
-            certificate.setId(keyHolder.getKey().intValue());
+    public Certificate add(Certificate certificate) {
+        try {
+            addTagIfNotExists(certificate.getTagList());
+            entityManager.persist(certificate);
+        } catch (PersistenceException exception) {
+            throw new ResourceAlreadyExistsException(certificate.getName());
         }
+        return certificate;
     }
 
     @Override
     public Certificate get(int id) {
-        return jdbcTemplate.query(SELECT_FROM_CERTIFICATE_WHERE_ID, new CertificateMapper(), id)
-                .stream().findAny().orElseThrow(() -> new ResourceNotFoundException(id));
+        Optional<Certificate> certificate = Optional.ofNullable(entityManager.find(Certificate.class, id));
+        return certificate.orElseThrow(() -> new ResourceNotFoundException(id));
     }
 
     @Override
-    public List<Certificate> get(OrderDTO orderDTO, SearchDTO searchDTO) {
-        CertificateUpdateParameters getParameters = getSQLRequest.create(orderDTO, searchDTO);
-        return jdbcTemplate.query(getParameters.getSqlRequest(),
-                new CertificateMapper(), getParameters.getParameters().toArray());
+    public List<Certificate> get(Page page, Sort sort, Search search) {
```

Comment:
I think it's better to split this big method into smaller ones

Reply:
done